### ISSUE TITLE
Forward simulations error 

### ISSUE THREAD

**Issue by DEHewitt** [DEHewitt/portunid_particle_tracking#23]
Katana outputs following error when attempting to run most recent push of "gmc_forward_bran205.py". Any ideas @HaydenSchilling 

```
WARNING: Casting lon data to np.float32
WARNING: Casting lat data to np.float32
WARNING: Casting depth data to np.float32
WARNING: Casting lon data to np.float32
WARNING: Casting lat data to np.float32
WARNING: Casting depth data to np.float32
WARNING: Casting time data to np.float64
WARNING: Casting time data to np.float64
WARNING: Casting field data to np.float32
WARNING: Casting field data to np.float32
INFO: Compiled random ==> /scratch/pbs.637034[17].kman.restech.unsw.edu.au/parcels-15278054/parcels_random_f4895468-d13a-46db-9024-bc00b98b4cf9.so
INFO: Compiled random ==> /scratch/pbs.637034[17].kman.restech.unsw.edu.au/parcels-15278054/parcels_random_206b3751-fa08-469e-9148-a44d82733a26.so
Traceback (most recent call last):
  File "gmc_forward_bran2015.py", line 176, in <module>
    pset = ParticleSet.from_list(fieldset, pclass=SampleParticle, time=start_time, lon=lon, lat=lat, repeatdt=repeatdt)
  File "/home/z5278054/mypython3env/lib/python3.6/site-packages/parcels/particleset.py", line 203, in from_list
    return cls(fieldset=fieldset, pclass=pclass, lon=lon, lat=lat, depth=depth, time=time, repeatdt=repeatdt, lonlatdepth_dtype=lonlatdepth_dtype, **kwargs)
  File "/home/z5278054/mypython3env/lib/python3.6/site-packages/parcels/particleset.py", line 136, in __init__
    if time[0] and not np.allclose(time, time[0]):
IndexError: index 0 is out of bounds for axis 0 with size 0
/apps/python/3.6.5/lib/python3.6/site-packages/sklearn/cluster/k_means_.py:971: ConvergenceWarning: Number of distinct clusters (1) found smaller than n_clusters (2). Possibly due to duplicate points in X.
  return_n_iter=True)
INFO: Compiled SampleParticleSampleInitialAdvectionRK4SampleAgeSampleTempSampleBathySampleSaltSampleDistanceBrownianMotion2D ==> /scratch/pbs.637034[17].kman.restech.unsw.edu.au/parcels-15278054/2c48e4cac14908e349272e26c5c817f0_0.so
INFO: Unable to locate chunking hints from dask, thus estimating the max. chunk size heuristically. Please consider defining the 'chunk-size' for 'array' in your local dask configuration file (see http://oceanparcels.org/faq.html#field_chunking_config and https://docs.dask.org).
  0% |                                                                        |
=>> PBS: job killed: walltime 43223 exceeded limit 43140
```

**Comment by jaseeverett** [comment]
I don't think the Parcels errors are fatal. The key error is from PBS
```=>> PBS: job killed: walltime 43223 exceeded limit 43140```
You ran out of time. Check the wall time you set?

**Comment by DEHewitt** [comment]
Thanks @jaseeverett - Hayden suggested the same thing. I'm doing a much reduced run now (3 particles per release) to test. I also realised I had an error in a line of my .pbs `#PBS -J 0-419` when it should be `#PBS -J 0-461`. Would this cause an error, or just mean some jobs wouldn't run?

**Comment by jaseeverett** [comment]
That would just mean some of the runs won't run.

**Comment by jaseeverett** [comment]
First thing to check it that you are at the maximum wall-time for the short queue (11:59:59) and then if you are still running out of time, I suggest submitting two pbs jobs. One with `#PBS -J 0-230` and one with `#PBS -J 231-461`.

Happy to help troubleshoot if you continue to have problems

**Comment by DEHewitt** [comment]
Thanks @jaseeverett. I would add that this code is adapted from backwards runs that worked and were much quicker (1-2 hours for 100 particles per day from 14 locations). I am running some smaller jobs now to try to troubleshoot.

**Comment by DEHewitt** [comment]
Same error message for a simulation with 3 particles released daily from 3 sites. Wall time request = 11:59:00. The same code runs backwards (npart = 100, daily, 14 locations) in 1-2 hours.

**Comment by jaseeverett** [comment]
Before we worry about why the simulation took so long, can you confirm that the simulation hit the wall time? ie 12 hours from when you submitted it to now?

**Comment by DEHewitt** [comment]
Yep, I submitted it at 2pm yesterday and the execution terminated at 2am.

**Comment by jaseeverett** [comment]
Does it run locally backwards without erroring? Release 1 particle a day backwards from 1 location for a few years. Then work up to see where it breaks. 

I have also tried turning off all the kernels and then re-adding them 1 by 1 to work out what breaks.

I can have a go at this tomorrow if you point me to the simulation file on Github.

**Comment by DEHewitt** [comment]
It has been a while since I've run the backwards locally - but my intuition is yes. Thanks for the tips. It won't be until next week sometime that I get to this, so in the meantime if you're happy to the files are "gmc_forward_bran2015.py" and "gmc_back_bran2015.py" in the 'Simulations' folder. I can ask Duncan to give you access to the directory (in my scratch) where the BRAN files are if needed.

**Comment by jaseeverett** [comment]
No worries. I'll try and have a look. I probably will need scratch access to BRAN if you can arrange it.

Are you going back into UNSW at all? I am starting to go once a week/fortnight at the moment if you want to catch up. I'd love to chat about your project and how its all going.

**Comment by DEHewitt** [comment]
I'll email ResTech now. What is your zID?

I am planning to make a trip down shortly for a change of scenery/meet with Iain. I'll email you and try to coordinate a day once I have some possible dates figured out.

**Comment by jaseeverett** [comment]
z9902002

Thanks.

I’m looking now.

**Comment by DEHewitt** [comment]
Thanks. I've emailed Duncan (you're cc'd). Hopefully it won't take long.

**Comment by DEHewitt** [comment]
@HaydenSchilling @jaseeverett you both should have read and execute permission for my Bran scratch folder where the files are.

**Comment by DEHewitt** [comment]
/srv/scratch/z5278054/BRAN_AUS

**Comment by DEHewitt** [comment]
Disregard - still having issues...

**Comment by HaydenSchilling** [comment]
Update - I ran a scaled down version on my laptop fine and on katana as an interactive job (`array_ref = 0`) and both were successful. This leads me to think the problem is with the array jobs running rather than the parcels code.

**Comment by jaseeverett** [comment]
Thanks Hayden. 

I also ran a version on my laptop using the few data files in the GitHub repo. Both forward and backwards simulations ran for equal amounts of time, and resulted in equivalent file sizes.

One thing I did notice which may make the array jobs fail, is that you don’t define `end_time` in the backwards simulations.

In the forward simulations you replicate the `start_time` (From memory this for each particle you release?)
https://github.com/DEHewitt/portunid_particle_tracking/blob/f7440df76e580c9c70c51eab141f96e80e1f2d6d/Simulations/gmc_forward_bran2015.py#L172

But in the backward simulations you don't replicate `end_time`. Looking at my old files I think you need to? Perhaps the backward simulations only run once (due to the length of `end_time`) which is why they finish much faster than the forwards simulations?

**Comment by HaydenSchilling** [comment]
@jaseeverett Good spot but it does not appear to be the answer, I do now remember removing that line from the backwards on a while ago as it seemed to do nothing, I think Parcels was updated to not need it. I just tried another array job with a reduced number of particles and after 15 min the particles were still on 0% when they were all done on the non-array version after that time.

**Comment by HaydenSchilling** [comment]
I've run the exact same code as an array job and normal job (manually setting the ref_number. The normal job ran successfully in less than 5min while the array job just stalled (did not progress beyond 0%). I think we should ask Restech.

**Comment by jaseeverett** [comment]
The last thing I would check is that when you run the array job, that the array_id is successfully passed to python.

To do this, comment out the actual parcels run line, and `print` the array id. It should be output into the .OU file. Then you can check that its not a problem with the array_id being mangled for some reason (I've had this before where it got into python (from the .pbs script) as an `NA`.

But otherwise, yes check with restech

**Comment by HaydenSchilling** [comment]
Good idea, I can confirm the array_id is passed successfully but i did notice something else in the output. Everything is duplicated:

```
This run is for array number  0
This run is for array number  0
WARNING: Casting lon data to np.float32
WARNING: Casting lon data to np.float32
WARNING: Casting lat data to np.float32
WARNING: Casting lat data to np.float32
WARNING: Casting depth data to np.float32
WARNING: Casting depth data to np.float32
WARNING: Casting time data to np.float64
WARNING: Casting time data to np.float64
WARNING: Casting field data to np.float32
WARNING: Casting field data to np.float32
```

So I think we are calling the MPI wrong or something but as the backwards ones worked (? @DEHewitt ) this is strange.

**Comment by jaseeverett** [comment]
I suspect this is because you call 2 cpus and 2 processors in the pbs file?

https://github.com/DEHewitt/portunid_particle_tracking/blob/f7440df76e580c9c70c51eab141f96e80e1f2d6d/Simulations/gmc_back_bran2015.pbs#L3

I've never used MPI so I can't comment specifically.... But given you do this in both the backward and forwards simulations I don't know what is different.

**Comment by jaseeverett** [comment]
Another 2 ideas (!) 

1) Change the lat/lon to the middle of the ocean to ensure its not about particles getting stuck. I note that the starting location for the forward and backwards simulations are the same. I assume this is intentional? Check the particles aren't getting stuck against the coast?
@HaydenSchilling you've said the simulation runs manually, but are you actually getting reasonable output? Or does it just finish.

2) Get the backwards code which know works, and change the delta to run forward. Does it run now?

https://github.com/DEHewitt/portunid_particle_tracking/blob/f7440df76e580c9c70c51eab141f96e80e1f2d6d/Simulations/gmc_back_bran2015.py#L182-L187

**Comment by DEHewitt** [comment]
I've just checked output from the _backwards_ simulations and it looks sensible, e.g.:

- correct dates

- dispersal pathways consistent with general expectations (i.e. mostly coming from the north)

Calling 2 cpus amd 2 processors is left over from last time ResTech helped us with some MPI issues.

I'm currently running the backwards code again to (double) confirm it works and will then try your second suggestion @jaseeverett

**Comment by HaydenSchilling** [comment]
1. Yes the output looks reasonable and the particles are not getting stuck.

2. Backwards code runs fine in backwards mode, not in forward .I also changed the time in pset line to :
`pset = ParticleSet.from_list(fieldset, pclass=SampleParticle, time=start_time, lon=lon, lat=lat, repeatdt=repeatdt)`
as i thought it made sense to have the start date at the beginning if going forward.

**Comment by HaydenSchilling** [comment]
The backwards code runs OK in forwards mode if you change the `time=start_time` to `time=end_time` but then it is simulating in the wrong time period.

**Comment by jaseeverett** [comment]
@HaydenSchilling Why? If you changed `time=start_time` to `time=end_time` it should be right? (Apart from the small discrepancy in `end_time` and `start_time` defined in the two files?

**Comment by DEHewitt** [comment]
Could this be to do with BRAN2015 only having velocities up until March 2019, but we are asking it to simulate from September 2019 to March 2020 with
```
start_time = datetime(year_array[array_ref], 8, 1) # year, month, day
end_time = datetime(year_array[array_ref]+1, 4, 30) # year, month, day
```
I think that the duplication is just the concatenation of messages that each cpu generates.

**Comment by jaseeverett** [comment]
If the backwards code runs ok in forwards mode, I suspect there is something weird with the actual py file?

**Comment by HaydenSchilling** [comment]
@jaseeverett I have no idea what the difference is. By changing the time command in the pset function from start_time to end_time I can swap between running and not running. It looks like the only place those variables are defined is the bit Dan mentioned above which both just generate a date string. I'll try and do some more experimenting.

@DEHewitt I am only using a 3 year array as a test (2009-2011) so it is not a problem running out of velocities in 2019 (yet)

**Comment by jaseeverett** [comment]
@HaydenSchilling Not sure what you mean about changing start_time to end_time, but this is what you must do to ensure the model runs correctly in backwards mode. You are essentially saying start at the `end_time` and run backwards. If you try and run backwards from `start_time` it will exit (or error) straight away because there are no more time steps prior to `start_time`

Happy to chat by phone later this arvo if that is easier....

**Comment by HaydenSchilling** [comment]
I was wrong it does appear to work both way (forwards and back just with highly variable times to completion - seems to get hung up on the removing previous file step sometimes).

```
# If output file already exists then remove it
if os.path.exists(out_file):
    os.remove(out_file)
```
Occasionally gives error of can't find the file and just pauses for a long time before moving onto the actual simulations

 I'll go back and look at the script that was written to go forwards.

**Comment by DEHewitt** [comment]
I got the backwards code (in this repo) to run forwards with the only changes being `dt=delta(minutes=5)` and `time=start_time`. The only other difference between the two is the way the array jobs are divided (i.e. by year and release location for forwards and only by year for backwards).

**Comment by DEHewitt** [comment]
Email sent to ResTech today (16/9/2020). You're both cc'd.

**Comment by HaydenSchilling** [comment]
The solution appears to be simple. The last line of the PBS needs to change from:
`mpirun -n 2 python3 gmc_forward_bran2015.py`
to
`mpirun -n 1 python3 gmc_forward_bran2015.py`

I have now run a forwards array job successfully after doing this (following Restech help)

**Comment by DEHewitt** [comment]
I've now run successful forwards simulations on Parcels v2.2 using the above fix. Output looks sensible. Testing wall time for a run with 1000 particles released daily during the spawning season (Sep-March; 242 days).

**Comment by DEHewitt** [comment]
Full scale simulation described above ran in <2 hours.
